### PR TITLE
1109 unix timestamp

### PR DESCRIPTION
**Pull request by dsawali**
Thank you for your contribution to Taquito.

Before submitting this PR, please make sure:

- [x] Your code builds cleanly without any errors or warnings
- [x] You have added unit tests
- [ ] You have added integration tests (if relevant/appropriate)
- [ ] All public methods or types have TypeDoc coverage with a complete description, and ideally an @example
- [ ] You have added or updated corresponding documentation
- [ ] If relevant, you have written a first draft summary describing the change for inclusion in Release Notes. 

## Release Note Draft Snippet

__If relevant, please write a summary of your change that will be suitable for
inclusion in the Release Notes for the next Taquito release.__

### CODE REVIEW

**Comment by dsawali** [comment]
This PR addresses [issue #1109](https://github.com/ecadlabs/taquito/issues/1109)

Added extra typecheck and added relevant tests to the timestamp class

File: packages/taquito-michelson-encoder/test/tokens/timestamp.spec.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import { TimestampToken } from '../../src/tokens/comparable/timestamp';
+
+describe('Timestamp token', () => {
+  let token: TimestampToken;
+
+  beforeEach(() => {
+    token = new TimestampToken({ prim: 'timestamp', args: [{ string: '1638554817'}], annots: [] }, 0, null as any);
```
Comment (roxaneletourneau) [inline packages/taquito-michelson-encoder/test/tokens/timestamp.spec.ts:7]: The args should be empty here as the token only represents the Michelson type, not the value.

File: packages/taquito-michelson-encoder/test/tokens/timestamp.spec.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  });
+
+  describe('Execute', () => {
+    it('Should return the RFC3339 time format when type is int and value is UNIX timestamp', () => {
+      const result = token.Execute({ int: 1638555790 } as any);
```
Comment (roxaneletourneau) [inline packages/taquito-michelson-encoder/test/tokens/timestamp.spec.ts:12]: { int: 1638555790 } should be { int: '1638555790' } and thus no need to cast to any.

File: packages/taquito-michelson-encoder/test/tokens/timestamp.spec.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      expect(result).toEqual('2021-12-03T18:23:10.000Z');
+    });
+
+    it('Should return the RFC3339 time format when type is string and value is UNIX timestamp', () => {
+      const result = token.Execute({ string: '1638555790' } as any);
```
Comment (roxaneletourneau) [inline packages/taquito-michelson-encoder/test/tokens/timestamp.spec.ts:17]: I don't think there is a need to cast to any here as the expected type of the parameter is `{ string?: string; int?: string }`.

File: packages/taquito-michelson-encoder/test/tokens/timestamp.spec.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+      expect(result).toEqual('2021-12-03T18:23:10.000Z')
+    });
+
+    it('Should return an error when value is not a valid Date or timestamp format', () => {
+      expect(() => {
+        token.Execute({ string: 'not valid' } as any);
```
Comment (roxaneletourneau) [inline packages/taquito-michelson-encoder/test/tokens/timestamp.spec.ts:23]: Same as the precedent comment, no need to cast to any

File: packages/taquito-michelson-encoder/src/tokens/comparable/timestamp.ts [hunk 5]
```diff
@@ -12,7 +12,10 @@ export class TimestampToken extends ComparableToken {
   }
 
   public Execute(val: { string?: string; int?: string }) {
-    if (val.string) {
+    if (val.string && /^\d{10}(\.\d)?$/.test(val.string)) {
+      const timestamp = +val.string;
+      return new Date(Number(timestamp) * 1000).toISOString();
```
Comment (roxaneletourneau) [inline packages/taquito-michelson-encoder/src/tokens/comparable/timestamp.ts:17]: "+" and "Number" have the same behavior of converting the string to a number. I suggest keeping only one of them.

File: packages/taquito-michelson-encoder/src/tokens/comparable/timestamp.ts [hunk 6]
```diff
@@ -12,7 +12,10 @@ export class TimestampToken extends ComparableToken {
   }
 
   public Execute(val: { string?: string; int?: string }) {
-    if (val.string) {
+    if (val.string && /^\d{10}(\.\d)?$/.test(val.string)) {
```
Comment (roxaneletourneau) [inline packages/taquito-michelson-encoder/src/tokens/comparable/timestamp.ts:15]: If we have a Unix timestamp that refers to the 2000s or prior, the regex will fail as the number of digits will be < 10.
Should the check be less restrictive?

**Review by roxaneletourneau — CHANGES_REQUESTED** [review]

File: packages/taquito-michelson-encoder/src/tokens/comparable/timestamp.ts (hunk 5 above)
Comment (dsawali) [reply packages/taquito-michelson-encoder/src/tokens/comparable/timestamp.ts:17]: you are correct, I missed the `Number()` casting

File: packages/taquito-michelson-encoder/src/tokens/comparable/timestamp.ts (hunk 6 above)
Comment (dsawali) [reply packages/taquito-michelson-encoder/src/tokens/comparable/timestamp.ts:15]: Good point, I can make it less restrictive. Do you think 9-10 digits is enough? or should we be making a larger range like 8-10?

File: packages/taquito-michelson-encoder/test/tokens/timestamp.spec.ts (hunk 1 above)
Comment (dsawali) [reply packages/taquito-michelson-encoder/test/tokens/timestamp.spec.ts:7]: oops nice catch, I meant them to be empty.

File: packages/taquito-michelson-encoder/src/tokens/comparable/timestamp.ts (hunk 6 above)
Comment (roxaneletourneau) [reply packages/taquito-michelson-encoder/src/tokens/comparable/timestamp.ts:15]: What do you think about just validating that the string is made of digits regardless of the length?

Comment (dsawali) [reply packages/taquito-michelson-encoder/src/tokens/comparable/timestamp.ts:15]: we can definitely do that too, I can make that change then

**Review by roxaneletourneau — APPROVED** [review]
Great work :)

**Pull request merged**